### PR TITLE
Add Email flow/API UT

### LINKED ISSUE #94: (Email) Basic flow unit test

### PR DESCRIPTION
**Pull request by DingmaomaoBJTU**
Add basic flow UT for Email Skill.

fixes #94

### CODE REVIEW

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/EmailBotTestBase.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
...pcGFkZHIiOiIxNjcuMjIwLjIzMi44MyIsIm5hbWUiOiJRaW9uZyBXdSIsIm9pZCI6ImI2OGQ1ZGNkLTZhYjYtNGZjMC05NmNmLTZhYmZkNmFjZjhkZiIsIm9ucHJlbV9zaWQiOiJTLTEtNS0yMS0yMTQ2NzczMDg1LTkwMzM2MzI4NS03MTkzNDQ3MDctMTc4MzAzMSIsInBsYXRmIjoiMyIsInB1aWQiOiIxMDAzN0ZGRTg4RjUzMzJCIiwic2NwIjoiQ2FsZW5kYXJzLlJlYWRXcml0ZSBNYWlsLlJlYWQgTWFpbC5TZW5kIE5vdGVzLlJlYWRXcml0ZSBQZW9wbGUuUmVhZCBVc2VyLlJlYWQgcHJvZmlsZSBvcGVuaWQgZW1haWwiLCJzaWduaW5fc3RhdGUiOlsiZHZjX21uZ2QiLCJkdmNfY21wIiwiZHZjX2RtamQiLCJrbXNpIl0sInN1YiI6Ik9Pci11VzlOdDZKMUNyMS1aZ2tIQ2w5UnQ1WG9Xdk5RNW1LeXYwX2dDUzgiLCJ0aWQiOiI3MmY5ODhiZi04NmYxLTQxYWYtOTFhYi0yZDdjZDAxMWRiNDciLCJ1bmlxdWVfbmFtZSI6InFpb3d1QG1pY3Jvc29mdC5jb20iLCJ1cG4iOiJxaW93dUBtaWNyb3NvZnQuY29tIiwidXRpIjoiYWN4MXpCZ0ZIVXlwZnFERjNtaFpBQSIsInZlciI6IjEuMCIsInhtc19zdCI6eyJzdWIiOiIxbVpyb0VBdTItWTNDcEFUWE4xOGtfLWhzWEt4eEx0MXQ2a3RVdzl3a1NzIn0sInhtc190Y2R0IjoxMjg5MjQxNTQ3fQ.p9Kzt4pER50GILElt3ZaNuXvLze9bXZiSTy37HiBZ5S_D0e9D_OsCWrvYEc7X0tZU06qWl6_GMqOLvBEpKZUakxz1WS2Ik0GBv05CGpGGq0qT7on49HSLzCIBGl3k9d5qKk2C1kgG9SxOkwYgZCFr4oze1gJgcFIM6YRrp55z5nDbZbEgyufdO4bnpZRqcjfWWpI3Fd_-aSEZuhEB6w_P2QUYhk6KMnYYRdYvV_LBWJzh3peTQd-xbj1jT1YqrXz-uef4ScFsaBRycLyZ8xS4UQDf-zbTXH6cM_waXbpd-7SxTzhNxAoyLX-qp0ELjPgU9iXx_Nq5XCCUWtFncPu1w";
```
Comment (lauren-mills) [inline solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/EmailBotTestBase.cs]: We aren't able to store tokens in code like this for security reasons.

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/appsettings.json [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    },
+    "Console": {
+      "LogLevel": {
+        "Default": "Warning"
+      }
+    }
+  },
+  // todo: set value
+  "BotId": "IPAEmailBotRegistration",
+  "MicrosoftAppId": "c4ecdfd8-4e9b-4190-a9ec-0c5690845314",
```
Comment (lauren-mills) [inline solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/appsettings.json]: These keys and secrets cannot be stored in the repo for security reasons.

**Review by lauren-mills — CHANGES_REQUESTED** [review]

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/EmailBotTestBase.cs (hunk 1 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/EmailBotTestBase.cs]: Sorry I forget this key. Currently we use fake graph service, so there is no need to use graph key. Will remove that~

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/appsettings.json (hunk 2 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/appsettings.json]: Yes, will remove all information in this file. Thanks!

**Review by lauren-mills — APPROVED** [review]

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/EmailSkillTest.csproj [hunk 3]
```diff
@@ -0,0 +1,49 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
+
+    <IsPackable>false</IsPackable>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
+    <CodeAnalysisRuleSet>..\..\..\samples.ruleset</CodeAnalysisRuleSet>
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/EmailSkillTest.csproj]: VirtualAssistant.ruleset

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/EmailSkillTest.csproj [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+    <CodeAnalysisRuleSet>..\..\..\samples.ruleset</CodeAnalysisRuleSet>
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/EmailSkillTest.csproj]: VirtualAssistant.ruleset. it's under the csharp folder

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ForwardEmailFlowTests.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    public class ForwardEmailFlowTests : EmailBotTestBase
+    {
+        [TestInitialize]
+        public override void Initialize()
+        {
+            base.Initialize();
+        }
+
+        [TestMethod]
+        public async Task EndToEnd()
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ForwardEmailFlowTests.cs]: Prefer better naming: Test_NotSendingEmail

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ForwardEmailFlowTests.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                .AssertReplyOneOf(this.CollectEmailContentMessage())
+                .Send("TestContent")
+                .AssertReply(this.AssertComfirmBeforeSendingPrompt())
+                .Send("No")
+                .AssertReplyOneOf(this.NotSendingMessage())
+                .StartTestAsync();
+        }
+
+        [TestMethod]
+        public async Task TestSending()
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ForwardEmailFlowTests.cs]: Prefer better naming: Test_SendingEmail

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ForwardEmailFlowTests.cs (hunk 5 above)
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ForwardEmailFlowTests.cs]: The test is not passing because it's missing EmailSkillTest.bot file. Did you forget to include it?

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ReplyFlowTests.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    public class ReplyFlowTests : EmailBotTestBase
+    {
+        [TestInitialize]
+        public override void Initialize()
+        {
+            base.Initialize();
+        }
+
+        [TestMethod]
+        public async Task EndToEnd()
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ReplyFlowTests.cs]: Test_NotSending.

Apply the naming change to all tests accordingly.

File: solutions/Virtual-Assistant/src/csharp/skills/test/TestFramework/TestFramework.csproj [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    <IsPackable>false</IsPackable>
+
+    <RootNamespace>TestFramework</RootNamespace>
+
+    <AssemblyName>TestFramework</AssemblyName>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
+    <CodeAnalysisRuleSet>..\..\..\samples.ruleset</CodeAnalysisRuleSet>
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/test/TestFramework/TestFramework.csproj]: VirtualAssistant.ruleset

File: solutions/Virtual-Assistant/src/csharp/skills/test/TestFramework/TestFramework.csproj [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+    <CodeAnalysisRuleSet>..\..\..\samples.ruleset</CodeAnalysisRuleSet>
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/test/TestFramework/TestFramework.csproj]: VirtualAssistant.ruleset

File: solutions/Virtual-Assistant/src/csharp/skills/test/TestFramework/TestFramework.csproj [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <None Remove="stylecop.json" />
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/test/TestFramework/TestFramework.csproj]: is this file needed?

File: solutions/Virtual-Assistant/src/csharp/skills/test/TestFramework/TestFramework.csproj [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+  </ItemGroup>
+
+  <ItemGroup>
+    <AdditionalFiles Include="stylecop.json" />
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/test/TestFramework/TestFramework.csproj]: is this file needed? and also you've got a Remove node on line 22 for the exact same file.

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs [hunk 12]
```diff
@@ -100,6 +100,10 @@ public async Task<DialogTurnResult> GetAuthToken(WaterfallStepContext sc, Cancel
                     // Wait for the tokens/response event
                     return await sc.PromptAsync(SkillModeAuth, new PromptOptions());
                 }
+                else if (sc.Context.Activity.ChannelId == "test")
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: this feels weird. did you just add this branch for the purpose of testing?

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/MailService.cs [hunk 13]
```diff
@@ -251,6 +247,51 @@ public async Task<Message> UpdateMessage(Message updatedMessage)
...
+        /// </summary>
+        /// <param name="id">Message id.</param>
+        public async Task DeleteMessage(string id)
+        {
+            await this.DeleteMessageFromGraphAsync(id);
+        }
+
+        protected virtual async Task<IMailFolderMessagesCollectionPage> GetMessagesFromGraphAsync(List<QueryOption> optionList)
+        {
+            IMailFolderMessagesCollectionPage messages = optionList.Count != 0 ? await this._graphClient.Me.MailFolders.Inbox.Messages.Request(optionList).GetAsync() : await this._graphClient.Me.MailFolders.Inbox.Messages.Request().GetAsync();
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/MailService.cs:265]: split this to multiple lines to make it more readable

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/API/MailServiceTests.cs [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    [TestClass]
+    public class MailServiceTests
+    {
+        private static FakeMailService mailService;
+        private static string testEmailAddress = "test@test.com";
+
+        [ClassInitialize]
+        public static void ClassInit(TestContext context)
+        {
+            mailService = new FakeMailService("", timeZoneInfo: TimeZoneInfo.Local);
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/API/MailServiceTests.cs:20]: string.empty.

do this for the entire PR accordingly

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/API/MailServiceTests.cs [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+            {
+                EmailAddress = new EmailAddress(),
+            };
+            recipient.EmailAddress.Address = testEmailAddress;
+            recipient.EmailAddress.Name = "Test Test";
+
+            List<Recipient> recipientList = new List<Recipient>();
+            recipientList.Add(recipient);
+
+            await mailService.SendMessage("test content", "test subject", recipientList);
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/API/MailServiceTests.cs:37]: mailService is an instance of FakeMailService. So we're not really testing the MailService class but testing the FakeMailService class, although i understand the function SendMessage is inherited from MailService implementation.

I think we should do this

1. create a static class called GraphServiceClientInitializer
2. create a static function in GraphServiceClientInitializer called CreateInstance and put this in:
        return  new GraphServiceClient(
                new DelegateAuthenticationProvider(
                    async (requestMessage) =>
                    {
                        // Append the access token to the request.
                        requestMessage.Headers.Authorization = new AuthenticationHeaderValue("bearer", accessToken);

                        // Get event times in the current time zone.
                        requestMessage.Headers.Add("Prefer", "outlook.timezone=\"" + _timeZoneInfo.Id + "\"");

                        await Task.CompletedTask;
                    }));
3. In MailService constructor, put in a third parameter with the type GraphServiceClient and name graphServiceClient and use that to assign to this._graphClient
4. In all places that MailService constructor is called, put in the third parameter as GraphServiceClientInitializer.CreateInstance()
5. In MailService implementation, extract all function that does non-graph related tasks into separate public functions.
5. Remove FakeMailService
6. Revert all "virtual" changes on MailService class.
7. Make mailService here an instance of MailService, instead of FakeMailService. pass in Null for the third parameter in the constructor
8. In the test, only test the non-graph related public functions. We're trying to make sure our code works, not to test the GraphClient code. 
8. For UserService, do the same thing since we refactored the code to initialize the GraphServiceClient out to the static class

This way we'll be testing the MailService class directly. The dependency is extracted out to a static helper class.

I understand this could be a little too much. Feel free to ping me on Teams or SFB so we can talk more in details

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ForwardEmailFlowTests.cs [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+                .AssertReplyOneOf(this.CollectRecipientsMessage())
+                .Send("TestName")
+                .AssertReply(this.ShowEmailList())
+                .AssertReply(this.AssertSelectOneOfTheMessage())
+                .Send("First")
+                .AssertReplyOneOf(this.CollectEmailContentMessage())
+                .Send("TestContent")
+                .AssertReply(this.AssertComfirmBeforeSendingPrompt())
+                .Send("No")
+                .AssertReplyOneOf(this.NotSendingMessage())
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ForwardEmailFlowTests.cs]: I feel like we shouldn't be using the Fake* services for E2E test because we kinda want this to be real :) I think it's better to put the E2E tests on hold until we figure out the token retrieval solution.

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/EmailSkillTest.csproj (hunk 4 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/EmailSkillTest.csproj]: Thanks! Applied~

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ForwardEmailFlowTests.cs (hunk 5 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ForwardEmailFlowTests.cs]: Yes, seems filtered by git ignore.  I will add it back. By the way, seems flow test have a dependency of LUIS key, is that OK?

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ReplyFlowTests.cs (hunk 7 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ReplyFlowTests.cs]: Sure, thanks!

File: solutions/Virtual-Assistant/src/csharp/skills/test/TestFramework/TestFramework.csproj (hunk 11 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/test/TestFramework/TestFramework.csproj]: Removed, that file is for stylecop header define.

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs (hunk 12 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: Yes, do you have better solutions for that?

File: solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ForwardEmailFlowTests.cs (hunk 16 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/test/EmailSkillTest/Flow/ForwardEmailFlowTests.cs]: As discuss offline, flow tests will work as component test. For end to end tests, will enable them when token retrieval is done.
